clang/test/SemaCXX/cxx20-ctad-type-alias.cpp

```diff
@@ -237,8 +237,17 @@ static_assert(__is_same(decltype(s.t), int));
 // explicit deduction guide.
 Foo(int) -> Foo<X>;
 AFoo s2{i};
-// FIXME: the type should be X because of the above explicit deduction guide.
-static_assert(__is_same(decltype(s2.t), int));
+static_assert(__is_same(decltype(s2.t), X));
+
+
+template<class T>
+using BFoo = AFoo<T>;
+static_assert(__is_same(decltype(BFoo(i).t), X));
+
+
+Foo(double) -> Foo<int>;
+static_assert(__is_same(decltype(AFoo(1.0).t), int));
+static_assert(__is_same(decltype(BFoo(1.0).t), int));
```

Comment:
Can we get more complicated test cases, w/ multiple template params and multiple constructor arguments etc e.g.: https://compiler-explorer.com/z/x8PhEqWeo

We should really be trying harder to stress test our implementation.

Does it fix this case: https://godbolt.org/z/b84rzhjjK if not we need to open a bug.

Reply:
> Does it fix this case: https://godbolt.org/z/b84rzhjjK if not we need to open a bug.

That test case doesn't look related to these changes to me, I think it's a different issue.

Reply:
> https://compiler-explorer.com/z/x8PhEqWeo

A simpler example: https://compiler-explorer.com/z/Tdx1jd1bM

I think Clang (and MSVC) are correct in rejecting this example. The candidate deduction guide `(AFoo(T t, U) -> Foo<T, U>)` is not selected because the `is_deducible` type trait is not satisfied, as indicated in the diagnostic message.

It looks like a bug in gcc.

Reply:
> > Does it fix this case: https://godbolt.org/z/b84rzhjjK if not we need to open a bug.
> 
> That test case doesn't look related to these changes to me, I think it's a different issue.

This is not related to this patch. Filed #125821 for tracking (thanks for spotting it!)